File: backend/app/services/mission/geo.py
"""Lightweight geodesy helpers (standard library only).

The grid planner works in a local east/north tangent plane measured in **feet**
(US contractor standard, spec §14.2). For the parcel-scale areas this system
deals with (a few hundred feet across) a simple equirectangular projection about
the polygon centroid is more than accurate enough — sub-foot error over the
working area, well within the ±1 ft accuracy noted in spec §9.2.
"""
from __future__ import annotations

import math
from typing import List, Sequence, Tuple

# Mean Earth radius in feet (WGS84 authalic radius 6371008.8 m × 3.280839895).
EARTH_RADIUS_FT = 20902231.0
FEET_PER_METER = 3.280839895
METERS_PER_FOOT = 0.3048

Coord = Tuple[float, float]  # (lng, lat) — GeoJSON axis order


def feet_per_degree_lat() -> float:
    """Feet per degree of latitude (very nearly constant)."""
    return EARTH_RADIUS_FT * math.pi / 180.0


def feet_per_degree_lng(lat_deg: float) -> float:
    """Feet per degree of longitude at the given latitude."""
    return EARTH_RADIUS_FT * math.pi / 180.0 * math.cos(math.radians(lat_deg))


def ring_centroid(ring: Sequence[Coord]) -> Coord:
    """Area-weighted centroid of a closed polygon ring (lng, lat)."""
    pts = list(ring)
    if pts[0] != pts[-1]:
        pts = pts + [pts[0]]
    a = cx = cy = 0.0
    for (x0, y0), (x1, y1) in zip(pts[:-1], pts[1:]):
        cross = x0 * y1 - x1 * y0
        a += cross
        cx += (x0 + x1) * cross
        cy += (y0 + y1) * cross
    if abs(a) < 1e-12:  # degenerate — fall back to vertex mean
        n = len(pts) - 1
        return (sum(p[0] for p in pts[:-1]) / n, sum(p[1] for p in pts[:-1]) / n)
    a *= 0.5
    return (cx / (6 * a), cy / (6 * a))


class LocalPlane:
    """Convert between (lng, lat) degrees and local (east, north) feet."""

    def __init__(self, origin: Coord):
        self.origin_lng, self.origin_lat = origin
        self._fpd_lat = feet_per_degree_lat()
        self._fpd_lng = feet_per_degree_lng(self.origin_lat)

    def to_feet(self, coord: Coord) -> Tuple[float, float]:
        lng, lat = coord
        east = (lng - self.origin_lng) * self._fpd_lng
        north = (lat - self.origin_lat) * self._fpd_lat
        return east, north

    def to_lnglat(self, east: float, north: float) -> Coord:
        lng = self.origin_lng + east / self._fpd_lng
        lat = self.origin_lat + north / self._fpd_lat
        return (lng, lat)


def bounding_box_ft(plane: LocalPlane, ring: Sequence[Coord]) -> Tuple[float, float, float, float]:
    """Axis-aligned bounding box of a ring in local feet: (min_e, min_n, max_e, max_n)."""
    pts = [plane.to_feet(c) for c in ring]
    es = [p[0] for p in pts]
    ns = [p[1] for p in pts]
    return min(es), min(ns), max(es), max(ns)


def point_in_ring(east: float, north: float, ring_ft: Sequence[Tuple[float, float]]) -> bool:
    """Ray-casting point-in-polygon test in the local feet plane."""
    inside = False
    n = len(ring_ft)
    j = n - 1
    for i in range(n):
        xi, yi = ring_ft[i]
        xj, yj = ring_ft[j]
        if ((yi > north) != (yj > north)) and (
            east < (xj - xi) * (north - yi) / (yj - yi + 1e-15) + xi
        ):
            inside = not inside
        j = i
    return inside


def polygon_dimensions_ft(plane: LocalPlane, ring: Sequence[Coord]) -> Tuple[float, float]:
    """Width (E–W) and height (N–S) of a ring's bounding box, in feet."""
    min_e, min_n, max_e, max_n = bounding_box_ft(plane, ring)
    return (max_e - min_e, max_n - min_n)

File: pwa/src/lib/offline/geo.js
// Lightweight geodesy — browser port of backend app/services/mission/geo.py.
// Works in a local east/north tangent plane measured in FEET about the polygon
// centroid. Accurate to sub-foot over parcel-scale areas. Keeping this identical
// to the Python keeps offline (phone) and online (server) output in lock-step.

export const EARTH_RADIUS_FT = 20902231.0
export const FEET_PER_METER = 3.280839895
export const METERS_PER_FOOT = 0.3048

export function feetPerDegreeLat() {
  return (EARTH_RADIUS_FT * Math.PI) / 180.0
}

export function feetPerDegreeLng(latDeg) {
  return ((EARTH_RADIUS_FT * Math.PI) / 180.0) * Math.cos((latDeg * Math.PI) / 180.0)
}

// Area-weighted centroid of a closed ring of [lng, lat] points.
export function ringCentroid(ring) {
  const pts = ring.slice()
  if (pts[0][0] !== pts[pts.length - 1][0] || pts[0][1] !== pts[pts.length - 1][1]) {
    pts.push(pts[0])
  }
  let a = 0,
    cx = 0,
    cy = 0
  for (let i = 0; i < pts.length - 1; i++) {
    const [x0, y0] = pts[i]
    const [x1, y1] = pts[i + 1]
    const cross = x0 * y1 - x1 * y0
    a += cross
    cx += (x0 + x1) * cross
    cy += (y0 + y1) * cross
  }
  if (Math.abs(a) < 1e-12) {
    const n = pts.length - 1
    let sx = 0,
      sy = 0
    for (let i = 0; i < n; i++) {
      sx += pts[i][0]
      sy += pts[i][1]
    }
    return [sx / n, sy / n]
  }
  a *= 0.5
  return [cx / (6 * a), cy / (6 * a)]
}

export class LocalPlane {
  constructor([originLng, originLat]) {
    this.originLng = originLng
    this.originLat = originLat
    this.fpdLat = feetPerDegreeLat()
    this.fpdLng = feetPerDegreeLng(originLat)
  }
  toFeet([lng, lat]) {
    return [(lng - this.originLng) * this.fpdLng, (lat - this.originLat) * this.fpdLat]
  }
  toLngLat(east, north) {
    return [this.originLng + east / this.fpdLng, this.originLat + north / this.fpdLat]
  }
}

export function boundingBoxFt(plane, ring) {
  const pts = ring.map((c) => plane.toFeet(c))
  const es = pts.map((p) => p[0])
  const ns = pts.map((p) => p[1])
  return [Math.min(...es), Math.min(...ns), Math.max(...es), Math.max(...ns)]
}

// Ray-casting point-in-polygon test in degrees (lng/lat) — used for parcel
// containment against cached county GeoJSON.
export function pointInRing(lng, lat, ring) {
  let inside = false
  const n = ring.length
  let j = n - 1
  for (let i = 0; i < n; i++) {
    const xi = ring[i][0],
      yi = ring[i][1]
    const xj = ring[j][0],
      yj = ring[j][1]
    if (yi > lat !== yj > lat && lng < ((xj - xi) * (lat - yi)) / (yj - yi + 1e-15) + xi) {
      inside = !inside
    }
    j = i
  }
  return inside
}

export function polygonDimensionsFt(plane, ring) {
  const [minE, minN, maxE, maxN] = boundingBoxFt(plane, ring)
  return [maxE - minE, maxN - minN]
}

// Shoelace area in square feet (for the "measuring" readout).
export function polygonAreaSqft(ring) {
  const plane = new LocalPlane(ringCentroid(ring))
  const pts = ring.map((c) => plane.toFeet(c))
  let area = 0
  for (let i = 0; i < pts.length - 1; i++) {
    const [x0, y0] = pts[i]
    const [x1, y1] = pts[i + 1]
    area += x0 * y1 - x1 * y0
  }
  // close the ring if needed
  const [xl, yl] = pts[pts.length - 1]
  const [xf, yf] = pts[0]
  area += xl * yf - xf * yl
  return Math.abs(area) / 2.0
}

// Extract the outer ring from a GeoJSON Polygon / MultiPolygon / Feature.
export function outerRing(geojson) {
  let geom = geojson
  if (geom.type === 'Feature') geom = geom.geometry
  if (geom.type === 'FeatureCollection') geom = geom.features[0].geometry
  if (geom.type === 'Polygon') return geom.coordinates[0].map(([x, y]) => [+x, +y])
  if (geom.type === 'MultiPolygon') return geom.coordinates[0][0].map(([x, y]) => [+x, +y])
  throw new Error(`Unsupported geometry type: ${geom.type}`)
}
